README.md
# Deno Cricket API 🦕 🏏

[![Test Lint](https://github.com/mskian/deno-cricket-api/actions/workflows/test.yml/badge.svg)](https://github.com/mskian/deno-cricket-api/actions/workflows/test.yml)
[![Create Release](https://github.com/mskian/deno-cricket-api/actions/workflows/release.yml/badge.svg)](https://github.com/mskian/deno-cricket-api/actions/workflows/release.yml)
![Deno](https://img.shields.io/badge/Deno-464647?style=for-the-badge&logo=deno&logoColor=white)

Live Cricket Score JSON API - Build using Deno 🦕

## Prerequisites 📕

- Deno on your System <https://deno.land/#installation>
- Denon for Development Server and Live Changes -
  <https://github.com/denosaurs/denon>

## Methods Used 🔧

- oak middleware framework for Deno's native HTTP server -
  <https://github.com/oakserver/oak>
- cheerio for data Scrape - <https://github.com/mskian/deno-cheerio-js>
- Cors - <https://github.com/tajpouria/cors>
- API rate limit - <https://github.com/AdityaTD/oak-rate-limit>

## Testing ⚙

- Development

```sh
denon run --allow-net --allow-read server.ts

or

deno task dev
```

- Production

```sh
deno run --allow-net --allow-read server.ts

or

deno task start
```

## API URL 🔵

- Live Match - `http://localhost:3000/live`
- Get Match Score by ID - `http://localhost:3000/match/<MatchID>`

### Configuration

- Update live Match URL in `/data/config.json`
- Update API Rate limits - `/server.ts`
- Update CORS Headers to allow Specific websites to access the API -
  `/lib/router.ts`
- change the Port No on `server.ts` file

## Hosting Support and Setup 📦

You can Host this on your own server or use Deno Deploy just connect it via
GitHub 👇

- Fork this Project on your Github Repo
- Link your Forked Project Repo on Deno Deploy, and `main` Branch to deploy
- Next, Choose the deploy option to automatic and Link the `server.ts` file to
  Start the API Server
- Finally, Click the **Link** Button to Connect the repo for Continous auto
  deploy

![Deno Deploy](./images/deno-deploy.png)

**Note** : `server.ts` is the Main file to Start API server if you are Hosting
this on your Server don't Forget to install deno and denon (For Local Testing)

```sh
# Production Start
deno task start

or

deno run --allow-net --allow-read server.ts
```

- Download the Project from Release Tag:
  <https://github.com/mskian/deno-cricket-api/releases> or use git to clone the
  Project repo on your Server
- Create Systemd.service file to run the app forever
- Use Nginx reverse proxy for DENO api server

> self Host this API Service for 100% Uptime ⬆

- Personally, I recommended **Deno deploy** - they offer a Free plan too
- For More About Deno Deploy Refer - <https://deno.com/deploy>

## Todo 🗃

- [x] Random User Agent
- [ ] API Caching
- [ ] Display Live Match ID No's in seperate API router

## Other Live Score API Projects 🗂

- Free Live Cricket Score API (PHP) - <https://github.com/mskian/cricket-api>
- Live Cricket Score JSON API (Node.js) -
  <https://github.com/mskian/cricket-api-nodejs>

## Contributing

Your PR's are Welcome 💚

## Disclaimer ⚠

- This is not an Offical API from Cricbuzz - it's an Unofficial API
- This is for Education Purpose only - use at your own risk on Production Site

## LICENSE ☑

MIT


CHANGELOG.md
<!-- deno-fmt-ignore-file -->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## [0.1.4] - 2023-03-20

- Update packages

## [0.1.3] - 2022-08-28

- Fix rate limit modul warning and fix router id response error

## [0.1.2] - 2022-08-26

- Update Dependencies
- Replace `any` to `number` in Get Live Match Data via ID

## [0.1.1] - 2022-08-23

- Readme improvements
- Update UserAgents

## [0.1.0] - 2022-08-22

The first version.


server.ts
import {
  Application,
  isHttpError,
} from "https://deno.land/x/oak@v12.1.0/mod.ts";
import router from "./lib/router.ts";
import { errorData } from "./data/error.ts";
import { RateLimiter } from "https://deno.land/x/oak_rate_limit@v0.1.1/mod.ts";

const HOST = "localhost";
const PORT = 3000;

const rateLimit = RateLimiter({
  windowMs: 1 * 60 * 1000,
  max: 45,
  headers: true,
  message:
    `{"title":"Too many requests, please try again later after 60 seconds.","update":"Data Not Found","current":"Too many requests, please try again later.","batsman":"Data Not Found","batsmanrun":"Data Not Found","ballsfaced":"Data Not Found","fours":"Data Not Found","sixes":"Data Not Found","sr":"Data Not Found","batsmantwo":"Data Not Found","batsmantworun":"Data Not Found","batsmantwoballsfaced":"Data Not Found","batsmantwofours":"Data Not Found","batsmantwosixes":"Data Not Found","batsmantwosr":"Data Not Found","bowler":"Data Not Found","bowlerover":"Data Not Found","bowlerruns":"Data Not Found","bowlerwickets":"Data Not Found","bowlermaiden":"Data Not Found","bowlertwo":"Data Not Found","bowletworover":"Data Not Found","bowlertworuns":"Data Not Found","bowlertwowickets":"Data Not Found","bowlertwomaiden":"Data Not Found","partnership":"Data Not Found","recentballs":"Data Not Found","lastwicket":"Data Not Found","runrate":"Data Not Found","commentary":"Data Not Found"}`,
  statusCode: 429,
});

const app = new Application();

app.use(async (ctx, next) => {
  try {
    await next();
  } catch (err) {
    if (isHttpError(err)) {
      switch (err.status) {
        case 404:
          ctx.response.status = 404;
          ctx.response.body = errorData;
          break;
        default:
          ctx.response.status = 400;
          ctx.response.body = errorData;
      }
    } else {
      ctx.response.status = 500;
      ctx.response.body = errorData;
    }
  }
});

app.use(await rateLimit);
app.use(router.routes());
app.use(router.allowedMethods());

console.log(`Listening on port ${PORT} ...`);
await app.listen(`${HOST}:${PORT}`);


data/error.ts
export function errorData() {
  return {
    title: "Data Not Found",
    update: "Data Not Found",
    current: "Data Not Found",
    batsman: "Data Not Found",
    batsmanrun: "Data Not Found",
    ballsfaced: "Data Not Found",
    fours: "Data Not Found",
    sixes: "Data Not Found",
    sr: "Data Not Found",
    batsmantwo: "Data Not Found",
    batsmantworun: "Data Not Found",
    batsmantwoballsfaced: "Data Not Found",
    batsmantwofours: "Data Not Found",
    batsmantwosixes: "Data Not Found",
    batsmantwosr: "Data Not Found",
    bowler: "Data Not Found",
    bowlerover: "Data Not Found",
    bowlerruns: "Data Not Found",
    bowlerwickets: "Data Not Found",
    bowlermaiden: "Data Not Found",
    bowlertwo: "Data Not Found",
    bowletworover: "Data Not Found",
    bowlertworuns: "Data Not Found",
    bowlertwowickets: "Data Not Found",
    bowlertwomaiden: "Data Not Found",
    partnership: "Data Not Found",
    recentballs: "Data Not Found",
    lastwicket: "Data Not Found",
    runrate: "Data Not Found",
    commentary: "Data Not Found",
  };
}


data/match.ts
import { cheerio } from "https://deno.land/x/denocheerio@1.0.0/mod.ts";
import { getAgent } from "./randomagent.ts";

type scoreData = {
  title: string;
  update: string;
  current: string;
  batsman: string;
  batsmanrun: string;
  ballsfaced: string;
  fours: string;
  sixes: string;
  sr: string;
  batsmantwo: string;
  batsmantworun: string;
  batsmantwoballsfaced: string;
  batsmantwofours: string;
  batsmantwosixes: string;
  batsmantwosr: string;
  bowler: string;
  bowlerover: string;
  bowlerruns: string;
  bowlerwickets: string;
  bowlermaiden: string;
  bowlertwo: string;
  bowletworover: string;
  bowlertworuns: string;
  bowlertwowickets: string;
  bowlertwomaiden: string;
  partnership: string;
  recentballs: string;
  lastwicket: string;
  runrate: string;
  commentary: string;
};

export async function dailyMatch(matchId: number) {
  return await fetch(
    `https://m.cricbuzz.com/live-cricket-scores/${matchId}/`,
    {
      headers: new Headers({
        "User-agent": getAgent(),
      }),
    },
  )
    .then((result) => result.text())
    .then((html) => {
      const $ = cheerio.load(html);
      const title = $("h4.ui-header").text();
      const update = $("div.cbz-ui-status").text();
      const currentscore = $("span.ui-bat-team-scores").text();
      const batsman = $("span.bat-bowl-miniscore").eq(0).text();
      const batsmanrun = $('td[class="cbz-grid-table-fix "]').eq(6).text();
      const ballsfaced = $('span[style="font-weight:normal"]').eq(0).text();
      const fours = $('td[class="cbz-grid-table-fix "]').eq(7).text();
      const sixes = $('td[class="cbz-grid-table-fix "]').eq(8).text();
      const sr = $('td[class="cbz-grid-table-fix "]').eq(9).text();
      const batsmantwo = $('td[class="cbz-grid-table-fix "]').eq(10).text();
      const batsmantworun = $('td[class="cbz-grid-table-fix "]').eq(11).text();
      const batsmantwoballsfaced = $('span[style="font-weight:normal"]').eq(1)
        .text();
      const batsmantwofours = $('td[class="cbz-grid-table-fix "]').eq(12)
        .text();
      const batsmantwosixes = $('td[class="cbz-grid-table-fix "]').eq(16)
        .text();
      const batsmantwosr = $('td[class="cbz-grid-table-fix "]').eq(14).text();
      const bowler = $("span.bat-bowl-miniscore").eq(2).text();
      const bowlerover = $('td[class="cbz-grid-table-fix "]').eq(21).text();
      const bowlerruns = $('td[class="cbz-grid-table-fix "]').eq(23).text();
      const bowlerwickets = $('td[class="cbz-grid-table-fix "]').eq(24).text();
      const bowlermaiden = $('td[class="cbz-grid-table-fix "]').eq(22).text();
      const bowlertwo = $("span.bat-bowl-miniscore").eq(3).text();
      const bowletworover = $('td[class="cbz-grid-table-fix "]').eq(26).text();
      const bowlertworuns = $('td[class="cbz-grid-table-fix "]').eq(28).text();
      const bowlertwowickets = $('td[class="cbz-grid-table-fix "]').eq(29)
        .text();
      const bowlertwomaiden = $('td[class="cbz-grid-table-fix "]').eq(27)
        .text();
      const partnership = $("span[style='color:#333']").eq(0).text();
      const recentballs = $("span[style='color:#333']").eq(2).text();
      const lastwicket = $("span[style='color:#333']").eq(1).text();
      const runrate = $("span[class='crr']").eq(0).text();
      const commentary = $("p[class='commtext']").text();
      const livescore = {
        title: title || "Data Not Found",
        update: update || "Data Not Found",
        current: currentscore || "Data Not Found",
        batsman: batsman || "Data Not Found",
        batsmanrun: batsmanrun || "Data Not Found",
        ballsfaced: ballsfaced || "Data Not Found",
        fours: fours || "Data Not Found",
        sixes: sixes || "Data Not Found",
        sr: sr || "Data Not Found",
        batsmantwo: batsmantwo || "Data Not Found",
        batsmantworun: batsmantworun || "Data Not Found",
        batsmantwoballsfaced: batsmantwoballsfaced || "Data Not Found",
        batsmantwofours: batsmantwofours || "Data Not Found",
        batsmantwosixes: batsmantwosixes || "Data Not Found",
        batsmantwosr: batsmantwosr || "Data Not Found",
        bowler: bowler || "Data Not Found",
        bowlerover: bowlerover || "Data Not Found",
        bowlerruns: bowlerruns || "Data Not Found",
        bowlerwickets: bowlerwickets || "Data Not Found",
        bowlermaiden: bowlermaiden || "Data Not Found",
        bowlertwo: bowlertwo || "Data Not Found",
        bowletworover: bowletworover || "Data Not Found",
        bowlertworuns: bowlertworuns || "Data Not Found",
        bowlertwowickets: bowlertwowickets || "Data Not Found",
        bowlertwomaiden: bowlertwomaiden || "Data Not Found",
        partnership: partnership || "Data Not Found",
        recentballs: recentballs || "Data Not Found",
        lastwicket: lastwicket || "Data Not Found",
        runrate: runrate || "Data Not Found",
        commentary: commentary || "Data Not Found",
      };
      const score: scoreData = livescore;
      return score;
    }).catch((error) => {
      console.log(error);
    });
}


lib/controller.ts
import { livescore } from "../data/score.ts";
import { dailyMatch } from "../data/match.ts";
import { errorData } from "../data/error.ts";

// deno-lint-ignore no-explicit-any
const homePage = ({ response }: { response: any }) => {
  response.headers.set(
    "Access-Control-Allow-Headers",
    "Access-Control-Allow-Headers,Content-Type,Access-Control-Allow-Methods, Authorization, X-Requested-With",
  );
  response.headers.set("Access-Control-Allow-Methods", "GET");
  response.headers.set("X-Frame-Options", "DENY");
  response.headers.set("Strict-Transport-Security", "max-age=63072000");
  response.body = { "message": "Cricket API" };
};

// deno-lint-ignore no-explicit-any
const getScore = async ({ response }: { response: any }) => {
  const ms = Date.now();
  response.status = 200;
  response.headers.set("X-Response-Time", `${ms}ms`);
  response.headers.set(
    "Access-Control-Allow-Headers",
    "Access-Control-Allow-Headers,Content-Type,Access-Control-Allow-Methods, Authorization, X-Requested-With",
  );
  response.headers.set("Access-Control-Allow-Methods", "GET");
  response.headers.set("X-Frame-Options", "DENY");
  response.headers.set("Strict-Transport-Security", "max-age=63072000");
  response.body = await livescore();
};

const matchScore = async (
  // deno-lint-ignore no-explicit-any
  { params, response }: { params: { id: any }; response: any },
) => {
  const currentMatch = params.id;
  const ms = Date.now();
  response.status = 200;
  response.headers.set("X-Response-Time", `${ms}ms`);
  response.headers.set(
    "Access-Control-Allow-Headers",
    "Access-Control-Allow-Headers,Content-Type,Access-Control-Allow-Methods, Authorization, X-Requested-With",
  );
  response.headers.set("Access-Control-Allow-Methods", "GET");
  response.headers.set("X-Frame-Options", "DENY");
  response.headers.set("Strict-Transport-Security", "max-age=63072000");
  response.body = await dailyMatch(currentMatch);
};

// deno-lint-ignore no-explicit-any
const ErrorHandle = ({ response }: { response: any }) => {
  response.status = 404;
  response.headers.set(
    "Access-Control-Allow-Headers",
    "Access-Control-Allow-Headers,Content-Type,Access-Control-Allow-Methods, Authorization, X-Requested-With",
  );
  response.headers.set("Access-Control-Allow-Methods", "GET");
  response.headers.set("X-Frame-Options", "DENY");
  response.headers.set("Strict-Transport-Security", "max-age=63072000");
  response.body = errorData;
};

export { ErrorHandle, getScore, homePage, matchScore };


data/randomagent.ts
export function getAgent() {
  const randomAgents = [
    "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 8.0.0; Pixel 2 XL Build/OPD1.170816.004) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Mobile Safari/537.36",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1",
    "Mozilla/5.0 (Linux; Android 6.0.1; Moto G (4)) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 9; Redmi Note 5 Pro) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.127 Mobile Safari/537.36",
    "Mozilla/5.0 (Android 9; Mobile; rv:81.0) Gecko/81.0 Firefox/81.0",
    "Mozilla/5.0 (Android 11; Mobile; rv:68.0) Gecko/68.0 Firefox/81.0",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 10_15_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) FxiOS/28.0 Mobile/15E148 Safari/605.1.15",
    "Mozilla/5.0 (Linux; Android 10; SM-G970F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Mobile Safari/537.36 OPR/59.1.2926.54067",
    "Mozilla/5.0 (Linux; Android 10; SM-A205U) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Mobile Safari/537.36",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.45 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36 OPR/67.0.3575.97",
    "Mozilla/5.0 (X11; Linux x86_64; rv:22.0) Gecko/20100101 Firefox/22.0",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Safari/605.1.15",
    "Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-A515F) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/15.0 Chrome/90.0.4430.210 Mobile Safari/537.36",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Linux; Android 12; SM-S906N Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/80.0.3987.119 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 10; SM-G996U Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 10; SM-G980F Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/78.0.3904.96 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 9; SM-G973U Build/PPR1.180610.011) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 12; Pixel 6 Build/SD1A.210817.023; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/94.0.4606.71 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 11; Pixel 5 Build/RQ3A.210805.001.A1; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/92.0.4515.159 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 7.1.1; G8231 Build/41.2.A.0.219; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/59.0.3071.125 Mobile Safari/537.36",
    "Mozilla/5.0 (iPhone14,3; U; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/602.1.50 (KHTML, like Gecko) Version/10.0 Mobile/19A346 Safari/602.1",
    "Mozilla/5.0 (iPhone13,2; U; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/602.1.50 (KHTML, like Gecko) Version/10.0 Mobile/15E148 Safari/602.1",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 12_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1",
    "Mozilla/5.0 (Apple-iPhone7C2/1202.466; U; CPU like Mac OS X; en) AppleWebKit/420+ (KHTML, like Gecko) Version/3.0 Mobile/1A543 Safari/419.3",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36 OPR/90.0.4480.48",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 12_5_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36 OPR/90.0.4480.48",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36 OPR/90.0.4480.48",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36 Vivaldi/5.3.2679.68",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36 Vivaldi/5.3.2679.68",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 12.5; rv:103.0) Gecko/20100101 Firefox/103.0",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:103.0) Gecko/20100101 Firefox/103.0",
    "Mozilla/5.0 (Linux; Android 7.1.2; Redmi 4A) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Mobile Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.114 Safari/537.36 Edg/103.0.1264.62",
    "Mozilla/5.0 (Linux; Android 9; Redmi Note 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Mobile Safari/537.36",
    "Mozilla/5.0 (Linux; Android 12; SM-G998U Build/SP1A.210812.016; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/103.0.5060.71 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/375.1.0.28.111;]",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 15_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/19F77 [FBAN/FBIOS;FBDV/iPhone12,1;FBMD/iPhone;FBSN/iOS;FBSV/15.5;FBSS/2;FBID/phone;FBLC/en_US;FBOP/5;FBIA/FBIOS]",
  ];
  const shuffle = randomAgents[Math.floor(Math.random() * randomAgents.length)];
  return shuffle;
}


data/score.ts
import { cheerio } from "https://deno.land/x/denocheerio@1.0.0/mod.ts";
import match from "./config.json" assert { type: "json" };
import { getAgent } from "./randomagent.ts";

type scoreData = {
  title: string;
  update: string;
  current: string;
  batsman: string;
  batsmanrun: string;
  ballsfaced: string;
  fours: string;
  sixes: string;
  sr: string;
  batsmantwo: string;
  batsmantworun: string;
  batsmantwoballsfaced: string;
  batsmantwofours: string;
  batsmantwosixes: string;
  batsmantwosr: string;
  bowler: string;
  bowlerover: string;
  bowlerruns: string;
  bowlerwickets: string;
  bowlermaiden: string;
  bowlertwo: string;
  bowletworover: string;
  bowlertworuns: string;
  bowlertwowickets: string;
  bowlertwomaiden: string;
  partnership: string;
  recentballs: string;
  lastwicket: string;
  runrate: string;
  commentary: string;
};

export async function livescore() {
  return await fetch(
    `https://m.cricbuzz.com/live-cricket-scores/${match.matchId}/`,
    {
      headers: new Headers({
        "User-agent": getAgent(),
      }),
    },
  )
    .then((result) => result.text())
    .then((html) => {
      const $ = cheerio.load(html);
      const title = $("h4.ui-header").text();
      const update = $("div.cbz-ui-status").text();
      const currentscore = $("span.ui-bat-team-scores").text();
      const batsman = $("span.bat-bowl-miniscore").eq(0).text();
      const batsmanrun = $('td[class="cbz-grid-table-fix "]').eq(6).text();
      const ballsfaced = $('span[style="font-weight:normal"]').eq(0).text();
      const fours = $('td[class="cbz-grid-table-fix "]').eq(7).text();
      const sixes = $('td[class="cbz-grid-table-fix "]').eq(8).text();
      const sr = $('td[class="cbz-grid-table-fix "]').eq(9).text();
      const batsmantwo = $('td[class="cbz-grid-table-fix "]').eq(10).text();
      const batsmantworun = $('td[class="cbz-grid-table-fix "]').eq(11).text();
      const batsmantwoballsfaced = $('span[style="font-weight:normal"]').eq(1)
        .text();
      const batsmantwofours = $('td[class="cbz-grid-table-fix "]').eq(12)
        .text();
      const batsmantwosixes = $('td[class="cbz-grid-table-fix "]').eq(16)
        .text();
      const batsmantwosr = $('td[class="cbz-grid-table-fix "]').eq(14).text();
      const bowler = $("span.bat-bowl-miniscore").eq(2).text();
      const bowlerover = $('td[class="cbz-grid-table-fix "]').eq(21).text();
      const bowlerruns = $('td[class="cbz-grid-table-fix "]').eq(23).text();
      const bowlerwickets = $('td[class="cbz-grid-table-fix "]').eq(24).text();
      const bowlermaiden = $('td[class="cbz-grid-table-fix "]').eq(22).text();
      const bowlertwo = $("span.bat-bowl-miniscore").eq(3).text();
      const bowletworover = $('td[class="cbz-grid-table-fix "]').eq(26).text();
      const bowlertworuns = $('td[class="cbz-grid-table-fix "]').eq(28).text();
      const bowlertwowickets = $('td[class="cbz-grid-table-fix "]').eq(29)
        .text();
      const bowlertwomaiden = $('td[class="cbz-grid-table-fix "]').eq(27)
        .text();
      const partnership = $("span[style='color:#333']").eq(0).text();
      const recentballs = $("span[style='color:#333']").eq(2).text();
      const lastwicket = $("span[style='color:#333']").eq(1).text();
      const runrate = $("span[class='crr']").eq(0).text();
      const commentary = $("p[class='commtext']").text();
      const livescore = {
        title: title || "Data Not Found",
        update: update || "Data Not Found",
        current: currentscore || "Data Not Found",
        batsman: batsman || "Data Not Found",
        batsmanrun: batsmanrun || "Data Not Found",
        ballsfaced: ballsfaced || "Data Not Found",
        fours: fours || "Data Not Found",
        sixes: sixes || "Data Not Found",
        sr: sr || "Data Not Found",
        batsmantwo: batsmantwo || "Data Not Found",
        batsmantworun: batsmantworun || "Data Not Found",
        batsmantwoballsfaced: batsmantwoballsfaced || "Data Not Found",
        batsmantwofours: batsmantwofours || "Data Not Found",
        batsmantwosixes: batsmantwosixes || "Data Not Found",
        batsmantwosr: batsmantwosr || "Data Not Found",
        bowler: bowler || "Data Not Found",
        bowlerover: bowlerover || "Data Not Found",
        bowlerruns: bowlerruns || "Data Not Found",
        bowlerwickets: bowlerwickets || "Data Not Found",
        bowlermaiden: bowlermaiden || "Data Not Found",
        bowlertwo: bowlertwo || "Data Not Found",
        bowletworover: bowletworover || "Data Not Found",
        bowlertworuns: bowlertworuns || "Data Not Found",
        bowlertwowickets: bowlertwowickets || "Data Not Found",
        bowlertwomaiden: bowlertwomaiden || "Data Not Found",
        partnership: partnership || "Data Not Found",
        recentballs: recentballs || "Data Not Found",
        lastwicket: lastwicket || "Data Not Found",
        runrate: runrate || "Data Not Found",
        commentary: commentary || "Data Not Found",
      };
      const score: scoreData = livescore;
      return score;
    }).catch((error) => {
      console.log(error);
    });
}
